File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P235_LowestCommonAncesterOfABinaryTree.java
package com.dsa.leetcode.labuladong.binary_tree;

public class P235_LowestCommonAncesterOfABinaryTree {

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int big = Math.max(p.val, q.val);
        int small = Math.min(p.val, q.val);
        return find(root, small, big);
    }

    private TreeNode find(TreeNode root, int small, int big) {
        if (root == null) {
            return null;
        }
        if (root.val < small) {
            return find(root.right, small, big);
        }
        if (root.val > big) {
            return find(root.left, small, big);
        }
        return root;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P235_LowestCommonAncesterOfABinaryTree.py
# author: tobebetter9527
# since: 2024/4/22 18:15
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        big = max(p.val, q.val)
        small = min(p.val, q.val)
        return self.find(root, small, big)

    def find(self, root, small, big):
        if not root:
            return None
        if root.val < small:
            return self.find(root.right, small, big)
        if root.val > big:
            return self.find(root.left, small, big)
        return root
